libraries/chain/include/eosio/chain/trace.hpp

```diff
@@ -65,6 +65,35 @@ namespace eosio { namespace chain {
       std::exception_ptr                         except_ptr;
    };
 
+   struct ram_trace {
```

Comment:
I removed `is_generic` but kept `generic_ram_trace` for now (see other comment about that).

I thought that `generic_ram_trace` would be easier for you guys to use and I could take care about the `action_id` and the rest, specially if not available where calls for RAM I made, I would make those changes at the same time.

Reply:
Here is my preferred way of handling it:
```cpp
struct ram_trace {
   uint32_t     action_id  = 0;
   const char*  event_id   = "generic";
   const char*  family     = "generic";
   const char*  operation  = "generic";
   const char*  legacy_tag = "generic";
};
```

And that is all that is needed.

Then it can be used throughout the code like (for example):
```cpp
update_db_usage(payer, config::billable_size_v<table_id_object>, ram_trace(get_action_id(), event_id.c_str(), "table", "add", "create_table"));
```
and
```cpp
      if (auto dm_logger = _get_deep_mind_logger()) {
         fc_dlog(*dm_logger, "RAM_OP ${action_id} ${event_id} ${family} ${operation} ${legacy_tag} ${payer} ${new_usage} ${delta}",
            ("action_id", trace.action_id)
            ("event_id", trace.event_id)
            ("family", trace.family)
            ("operation", trace.operation)
            ("legacy_tag", trace.legacy_tag)
            ("payer", account)
            ("new_usage", u.ram_usage)
            ("delta", ram_delta)
         );
      }
```
And we could use it for a generic RAM event like so:
```cpp
ram_trace{get_action_id()}
```

But I can understand if you want to make the generic one explicit. Also, I imagine you would like to avoid making changes everywhere you currently use `ram_trace` and `generic_ram_trace`. So how about the following?
```cpp
struct ram_trace {
public:
   ram_trace(uint32_t action_id, const char* event_id, const char* family, const char* operation, const char* legacy_tag)
   :action_id(action_id),event_id(event_id),family(family),operation(operation),legacy_tag(legacy_tag)
   {}

   inline uint32_t    get_action_id() const { return action_id; }	
   inline const char* get_event_id() const { return event_id; }
   inline const char* get_family() const { return family; }
   inline const char* get_operation() const { return operation; }
   inline const char* get_legacy_tag() const { return legacy_tag; }

private:
   ram_trace(uint32_t action_id)
   :action_id(action_id)
   {}

   friend ram_trace generic_ram_trace(uint32_t);

private:
   uint32_t     action_id  = 0;
   const char*  event_id   = "generic";
   const char*  family     = "generic";
   const char*  operation  = "generic";
   const char*  legacy_tag = "generic";
};

ram_trace generic_ram_trace(uint32_t action_id = 0) {
   return {action_id};
}
```
Or better yet avoid the default argument for `generic_ram_trace` (but then that would require some changes to the `resource_limits_test.cpp` file.

Once again that is all you would need to add to the header file and nothing would need to be added to the `trace.cpp` file.

Reply:
I went with an hybrid between your two version. I forced `action_id` in `generic_ram_trace(uint32_t action_id)` and made the fields public without any getter.

Tell me if alright with you.

Reply:
That's fine with me. Make sure to clean up the commented out code.

Reply:
Yeah, oups. Pushed fixed version.